### PR TITLE
Entities related to Entities of the same type

### PR DESCRIPTION
**Pull request by benbosman**
## Description
This resolves a bug if the left type and the right type of a relationship is identical. That setup was causing the virtual metadata to be populated incorrectly and the counts to be defined incorrectly.

## Instructions for Reviewers
The only 2 changes for the virtual metadata are in dspace-api/src/main/java/org/dspace/content/RelationshipMetadataServiceImpl.java to verify not only on the type, but also on the item's UUID
The counts are solved by defining whether the left or right side should be verified

There are already many existing ITs which still succeed.
A new type of relationship has been defined in the test now where the left type and the right type of a relationship is identical

## Checklist
- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide)
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java [hunk 1]
```diff
@@ -105,60 +116,60 @@ public void setUp() throws Exception {
         context.turnOffAuthorisationSystem();
 
         parentCommunity = CommunityBuilder.createCommunity(context)
-                                          .withName("Parent Community")
-                                          .build();
+                .withName("Parent Community")
+                .build();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:120]: @benbosman : Just a quick note that this PR needs a rebase & cleanup.  It looks like nearly all the changes to `RelationshipRestRepositoryIT` are **space changes** (where the IDE seems to have made a lot of unrelated alignment changes).  If you could revert those space changes it'd make this PR easier to review (and it might even fall into the "1 approval" category).

**Comment by benbosman** [comment]
@tdonohue the branch has been rebased

File: dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java [hunk 2]
```diff
@@ -201,18 +201,24 @@ public int countRows(Context context) throws SQLException {
     }
 
     @Override
-    public int countByItemAndRelationshipType(Context context, Item item, RelationshipType relationshipType)
-            throws SQLException {
+    public int countByItemAndRelationshipType(Context context, Item item, RelationshipType relationshipType,
+                                              boolean isLeft) throws SQLException {
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java:205]: I don't specially like the approach of passing parameters that in my perspective describe the property of an object. In this particular case the object Relationship. The concept of Left and Right item are always subjective and relative to the person that defines it. I think a different approach that could be considered in the future is to probably considers an Item not for the location/position which is it, but if it's itself in a relation verification.

But I understand the reason for having this boolean here. And considering the required effort for a different approach, I'm ok with having it this way in this version.

**Review by paulo-graca — APPROVED** [review]
My comment goes in the way of a repeating speech that we need to take a second/third look at the solution we found for relationships. From the first moment it was presented it raised doubts and some discussion. We are entering in a relative or subjective realm to the person that defines the relationship. 
The presented solution addresses the issue, but by adopting it we are adding complexity to the system (this is not particular to this PR, it's related with the current Relationship model). I suggest to consider a new approach perhaps it a further release or version of DSpace.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java [hunk 3]
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
...
+        MetadataSchema metadataSchema = metadataSchemaService.find(context, "relation");
+        MetadataFieldBuilder.createMetadataField(context, metadataSchema, "isParentOrgUnitOf", null, null).build();
+        MetadataFieldBuilder.createMetadataField(context, metadataSchema, "isChildOrgUnitOf", null, null).build();
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+
+        context.restoreAuthSystemState();
+        // Here we create our first Relationship to the Publication to give it a dc.contributor.author virtual
+        // metadata field.
+        MvcResult mvcResult = getClient(adminToken).perform(post("/api/core/relationships")
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:2519]: This `post()` and all the others in this IT do not seem to be cleaning up after themselves.  Remember we should be using a try/finally, using an `.andDo()` to capture the ID of the created object (I suspect this code was just written prior to that best practice being established).  In any case, here's an example: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/ScriptRestRepositoryIT.java#L195-L207

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java [hunk 4]
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
...
+        MetadataSchema metadataSchema = metadataSchemaService.find(context, "relation");
+        MetadataFieldBuilder.createMetadataField(context, metadataSchema, "isParentOrgUnitOf", null, null).build();
+        MetadataFieldBuilder.createMetadataField(context, metadataSchema, "isChildOrgUnitOf", null, null).build();
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+
+        context.restoreAuthSystemState();
+        // Here we create our first Relationship to the Publication to give it a dc.contributor.author virtual
+        // metadata field.
+        getClient(adminToken).perform(post("/api/core/relationships")
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:2566]: Again, `post` should cleanup after itself

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java [hunk 5]
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
...
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:2622]: Again, `post` should cleanup after itself

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java [hunk 6]
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
...
+                                                   .TEXT_URI_LIST_VALUE))
+                                          .content(
+                                              "https://localhost:8080/server/api/core/items/" + orgUnit1
+                                                  .getID() + "\n" +
+                                                  "https://localhost:8080/server/api/core/items" +
+                                                  "/" + orgUnit2
+                                                  .getID()))
+                             .andExpect(status().isCreated());
+
+        getClient(adminToken).perform(post("/api/core/relationships")
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:2636]: Again, `post` should cleanup after itself

File: dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java (hunk 2 above)
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java:205]: I agree with @paulo-graca 's comment here.  It seems like this `isLeft` is needed, but it's an awkward implementation...the left/right concept is turning out to be quite confusing (and it is subjective).  I'm not sure what to replace it with though, as that would take more detailed discussion that would likely need to be delayed for post 7.0

**Review by tdonohue — CHANGES_REQUESTED** [review]
@benbosman : I gave this a code review today.  The main concepts are OK, but the tests need some minor cleanup (to ensure `post` calls are cleaning up content they create).  See more comments inline below.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java (hunk 2 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java:205]: The Left/Right contents is indeed not pre-defined, and just dependent on how the relation is configured. The concept of Left/Right was chosen to make it easier if you draw the relationship on paper.
It's not an easy concept to support any relationship in code, but feel free to create a ticket with suggestions to make this easier to understand.
For 7.0 it's indeed not realistic to adjust this

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java (hunk 3 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This has been updated

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java (hunk 4 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This has been updated

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java (hunk 5 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This has been updated

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java (hunk 6 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java]: This has been updated

**Comment by benbosman** [comment]
@tdonohue Marie has completed the updates

She did notice HandleDAOImplTest stopped working, even though this PR doesn't touch any API tests. She confirmed that the test itself works. This most likely implies there's another test in the API which doesn't clean up data correctly. Due to the randomness in the tests, it suddenly overlapped. This has been temporarily bypassed in https://github.com/DSpace/DSpace/pull/2746/commits/d70e121e93a31dba899850aa399eca35aa7dbf82 because it's not easy to identify where the problem resides and it's out of scope for this PR.

I assume it's best to create a ticket for searching the TEST which creates the collection and doesn't remove it?

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now.  Thanks @benbosman and @MarieVerdonck !
